crates/subspace-farmer/src/cluster/controller/farms.rs

```diff
@@ -63,15 +84,18 @@ impl KnownFarmers {
 
     /// Return `false` if the farmer is unknown and initialization is required
     fn refresh(&mut self, farmer_id: ClusterFarmerId) -> bool {
-        self.known_farmers.iter_mut().any(|known_farmer| {
+        let mut is_new_farmer = false;
+        self.known_farmers.retain_mut(|known_farmer| {
             if known_farmer.farmer_id == farmer_id {
                 trace!(%farmer_id, "Updating last identification for farmer");
                 known_farmer.last_identification = Instant::now();
-                true
-            } else {
-                false
-            }
-        })
+                is_new_farmer = true;
+            };
+            // Remove the farmer when all farms are removed
+            !known_farmer.known_farms.is_empty()
+        });
+
+        is_new_farmer
```

Comment:
This doesn't feel like the right place to do the cleanup. If farmer doesn't have any farms, it should not be in the list of known farmers in the first place. Known farmer should be removed with last farm.